drivers/tsl2561: initial support

aabadie opened the pull request:
This PR adds support to the TSL2561 [illuminance sensor](https://cdn-shop.adafruit.com/datasheets/TSL2561.pdf) sensor.
Code is inspired from [Adafruit implementation](https://github.com/adafruit/TSL2561-Arduino-Library)

~~Caveat: I couldn't make it work without having to initialize the i2c bus for each read.~~
The interruption feature is not implemented but support has been added to saul.

aabadie commented:
I gave a try right after creating this PR and the sensor didn't work... Investigating.

PeterKietzmann commented:
@aabadie which board do you use for development? Probably #5460 is of interest?

aabadie commented:
> which board do you use for development? Probably #5460 is of interest?

I'm trying with a samr21-xpro board so indeed #5460 is interesting !

PeterKietzmann commented:
@aabadie did you fix the i2c issue or at least figure out if it's related the the samr21-xpro board or your driver?

aabadie commented:
> @aabadie did you fix the i2c issue or at least figure out if it's related the the samr21-xpro board or your driver?

I tested with an without and same problems. The conclusion is that it comes from my code (most likely) or from the sensor itself.

PeterKietzmann commented:
Thanks for the feedback. Please give a sign once PR isn't WIP any more.

PeterKietzmann commented:
Still WIP?

aabadie commented:
> Still WIP?

Basically yes, I have problems with I2C on this sensor and I don't know how to fix it.
I also noticed that `auto_init` introduces instability with this sensor because when calling `tsl2561_init` in a `main` function, the sensor is initialized twice. Thus I reverted the `auto_init` but the driver is still not working.
I also noticed that depending on the board, the driver was not working the same. I tested with samr21-xpro and nucleo-f401.
To be sure the sensor is working, I tested it with an arduino sketch and it worked fine.

PeterKietzmann commented:
@aabadie any news?

PeterKietzmann commented:
whoops...

aabadie commented:
> @aabadie any news?

Well, this is deep in my todo list. I wish I can have a look soon (within 2 or 3 weeks). I hope recent i2c changes with samr21 could help me as well (#5833, #5832).

aabadie commented:
Reviving this one as I could finally make it work :) Reviews are welcome.

PeterKietzmann commented on drivers/tsl2561/tsl2561.c:
```diff
@@ -0,0 +1,296 @@
+    DEBUG("[Info] I2C device: %d\n", dev->i2c_dev);
+    DEBUG("[Info] Address: %d\n", dev->addr);
+    switch(dev->gain) {
+    case TSL2561_GAIN_1X:
```
Would you indent the `case` please? I'm not 100% sure we have this in the coding conventions but it is "common" for RIOT.

aabadie replied:
Grepping for "switch" the indent of the case is a valid convention, will update

PeterKietzmann replied:
Indent here as well

PeterKietzmann commented on drivers/tsl2561/tsl2561.c:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    switch(dev->integration) {
+    case TSL2561_INTEGRATIONTIME_13MS:
```
Indentation here as well...

PeterKietzmann replied:
Here and following...

PeterKietzmann commented on drivers/tsl2561/tsl2561.c:
```diff
@@ -0,0 +1,296 @@
+    /* Acquire exclusive access */
+    i2c_acquire(dev->i2c_dev);
+
+#if ENABLE_DEBUG
```
Not needed wehen using `DEBUG()`

aabadie replied:
indeed, will update

PeterKietzmann commented on drivers/tsl2561/tsl2561.c:
```diff
@@ -0,0 +1,296 @@
+    uint8_t id;
+    i2c_read_reg(dev->i2c_dev, dev->addr,
+                 TSL2561_COMMAND_MODE | TSL2561_REGISTER_ID, &id);
+#if ENABLE_DEBUG
```
Not needed, or am I mistaken?

PeterKietzmann replied:
Here and following...

PeterKietzmann commented on drivers/tsl2561/tsl2561.c:
```diff
@@ -0,0 +1,296 @@
+#endif
+    if (id != TSL2561_ID ) {
+        DEBUG("[Error] not a TSL2561 sensor\n");
+        return -1;
```
This is something different than stated in the API documentation, right?
`-1 if given I2C is not enabled in board config`

PeterKietzmann commented on drivers/tsl2561/tsl2561_saul.c:
```diff
@@ -0,0 +1,45 @@
+ * @}
+ */
+
+#include <string.h>
```
Necessary here?

aabadie replied:
no, remove

PeterKietzmann commented on tests/driver_tsl2561/README.md:
```diff
@@ -0,0 +1,8 @@
+## Usage
+The application will initialize the TSL2561 sensor.
+
+After initialization, every 2 seconds, the application reads the measured illuminance
```
Looking at the code it looks rather like every 1 second.

gebart commented:
I think I have a Sparkfun break out board with a sensor named tsl25xx
something. I'll look for it and the adt7310 that also needs testing.

On Jan 16, 2017 4:49 PM, "Peter Kietzmann" <notifications@github.com> wrote:

> *@PeterKietzmann* commented on this pull request.
>
> @aabadie <https://github.com/aabadie> thanks for the implementation! The
> code looks fine except some minor issues. Unfortunately I don't have sensor
> to test this driver. An idea who has?
> And PR needs rebasing, by the way...
> ------------------------------
>
> In drivers/tsl2561/tsl2561.c
> <https://github.com/RIOT-OS/RIOT/pull/5570#pullrequestreview-16830361>:
>
> > +
> +#define ENABLE_DEBUG        (0)
> +#include "debug.h"
> +
> +/* internal helpers */
> +static void _enable(tsl2561_t *dev);
> +static void _disable(tsl2561_t *dev);
> +static void _read_data(tsl2561_t *dev, uint16_t *full, uint16_t *ir);
> +
> +#if ENABLE_DEBUG
> +static void _print_init_info(tsl2561_t *dev)
> +{
> +    DEBUG("[Info] I2C device: %d\n", dev->i2c_dev);
> +    DEBUG("[Info] Address: %d\n", dev->addr);
> +    switch(dev->gain) {
> +    case TSL2561_GAIN_1X:
>
> Would you indent the case please? I'm not 100% sure we have this in the
> coding conventions but it is "common" for RIOT.
> ------------------------------
>
> In drivers/tsl2561/tsl2561.c
> <https://github.com/RIOT-OS/RIOT/pull/5570#pullrequestreview-16830361>:
>
> > +    DEBUG("[Info] I2C device: %d\n", dev->i2c_dev);
> +    DEBUG("[Info] Address: %d\n", dev->addr);
> +    switch(dev->gain) {
> +    case TSL2561_GAIN_1X:
> +        DEBUG("[Info] Gain: 1X\n");
> +        break;
> +    case TSL2561_GAIN_16X:
> +        DEBUG("[Info] Gain: 16X\n");
> +        break;
> +    default:
> +        DEBUG("[Info] Invalid gain %d\n", dev->gain);
> +        break;
> +    }
> +
> +    switch(dev->integration) {
> +    case TSL2561_INTEGRATIONTIME_13MS:
>
> Indentation here as well...
> ------------------------------
>
> In drivers/tsl2561/tsl2561.c
> <https://github.com/RIOT-OS/RIOT/pull/5570#pullrequestreview-16830361>:
>
> > +#if ENABLE_DEBUG
> +    _print_init_info(dev);
> +#endif
> +
> +    /* Initialize I2C interface */
> +    if (i2c_init_master(dev->i2c_dev, I2C_SPEED_NORMAL)) {
> +        DEBUG("[Error] I2C device not enabled\n");
> +        return -1;
> +    }
> +
> +    DEBUG("[Info] I2C device initialized with success!\n");
> +
> +    /* Acquire exclusive access */
> +    i2c_acquire(dev->i2c_dev);
> +
> +#if ENABLE_DEBUG
>
> Not needed wehen using DEBUG()
> ------------------------------
>
> In drivers/tsl2561/tsl2561.c
> <https://github.com/RIOT-OS/RIOT/pull/5570#pullrequestreview-16830361>:
>
> > +    }
> +
> +    DEBUG("[Info] I2C device initialized with success!\n");
> +
> +    /* Acquire exclusive access */
> +    i2c_acquire(dev->i2c_dev);
> +
> +#if ENABLE_DEBUG
> +    DEBUG("[Info] Access acquired !\n");
> +#endif
> +
> +    /* Verify sensor ID */
> +    uint8_t id;
> +    i2c_read_reg(dev->i2c_dev, dev->addr,
> +                 TSL2561_COMMAND_MODE | TSL2561_REGISTER_ID, &id);
> +#if ENABLE_DEBUG
>
> Not needed, or am I mistaken?
> ------------------------------
>
> In drivers/tsl2561/tsl2561.c
> <https://github.com/RIOT-OS/RIOT/pull/5570#pullrequestreview-16830361>:
>
> > +    }
> +
> +    DEBUG("[Info] I2C device initialized with success!\n");
> +
> +    /* Acquire exclusive access */
> +    i2c_acquire(dev->i2c_dev);
> +
> +#if ENABLE_DEBUG
> +    DEBUG("[Info] Access acquired !\n");
> +#endif
> +
> +    /* Verify sensor ID */
> +    uint8_t id;
> +    i2c_read_reg(dev->i2c_dev, dev->addr,
> +                 TSL2561_COMMAND_MODE | TSL2561_REGISTER_ID, &id);
> +#if ENABLE_DEBUG
>
> Here and following...
> ------------------------------
>
> In drivers/tsl2561/tsl2561.c
> <https://github.com/RIOT-OS/RIOT/pull/5570#pullrequestreview-16830361>:
>
> > +    DEBUG("[Info] I2C device: %d\n", dev->i2c_dev);
> +    DEBUG("[Info] Address: %d\n", dev->addr);
> +    switch(dev->gain) {
> +    case TSL2561_GAIN_1X:
> +        DEBUG("[Info] Gain: 1X\n");
> +        break;
> +    case TSL2561_GAIN_16X:
> +        DEBUG("[Info] Gain: 16X\n");
> +        break;
> +    default:
> +        DEBUG("[Info] Invalid gain %d\n", dev->gain);
> +        break;
> +    }
> +
> +    switch(dev->integration) {
> +    case TSL2561_INTEGRATIONTIME_13MS:
>
> Here and following...
> ------------------------------
>
> In drivers/tsl2561/tsl2561.c
> <https://github.com/RIOT-OS/RIOT/pull/5570#pullrequestreview-16830361>:
>
> > +    i2c_acquire(dev->i2c_dev);
> +
> +#if ENABLE_DEBUG
> +    DEBUG("[Info] Access acquired !\n");
> +#endif
> +
> +    /* Verify sensor ID */
> +    uint8_t id;
> +    i2c_read_reg(dev->i2c_dev, dev->addr,
> +                 TSL2561_COMMAND_MODE | TSL2561_REGISTER_ID, &id);
> +#if ENABLE_DEBUG
> +    DEBUG("[Info] ID ? %d\n", id);
> +#endif
> +    if (id != TSL2561_ID ) {
> +        DEBUG("[Error] not a TSL2561 sensor\n");
> +        return -1;
>
> This is something different than stated in the API documentation, right?
> -1 if given I2C is not enabled in board config
> ------------------------------
>
> In drivers/tsl2561/tsl2561_saul.c
> <https://github.com/RIOT-OS/RIOT/pull/5570#pullrequestreview-16830361>:
>
> > + */
> +
> +/**
> + * @ingroup     drivers_tsl2561
> + * @{
> + *
> + * @file
> + * @brief       SAUL adaption for TSL2561 device
> + *
> + * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
> + * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
> + *
> + * @}
> + */
> +
> +#include <string.h>
>
> Necessary here?
> ------------------------------
>
> In tests/driver_tsl2561/README.md
> <https://github.com/RIOT-OS/RIOT/pull/5570#pullrequestreview-16830361>:
>
> > @@ -0,0 +1,8 @@
> +## About
> +This is a test application for the TSL2561 Lux sensor.
> +
> +## Usage
> +The application will initialize the TSL2561 sensor.
> +
> +After initialization, every 2 seconds, the application reads the measured illuminance
>
> Looking at the code it looks rather like every 1 second.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/5570#pullrequestreview-16830361>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATYQoEBlYNzuQpQWa-aCyTB_YnWZnFgks5rS5F5gaJpZM4I9s71>
> .
>

aabadie commented:
thanks ! I'll address the comments soon

PeterKietzmann commented:
Thanks @gebart! Yes, the adt7310 needs testing with the SPI rework as well and has priority for the release :-).

aabadie commented:
rebased

PeterKietzmann commented:
@aabadie will you address or answer my comments?

aabadie commented:
Ah yes, sorry @PeterKietzmann, I missed them. Will do later.

aabadie commented:
@PeterKietzmann, I addressed your comments and made a bit of cleanup in the code. I also fixed an issue with the values being read. Now it's all good : the returned values are very meaningful (around 400lx in my office).
I also applied the changed from #6385.

PeterKietzmann commented on drivers/tsl2561/tsl2561.c:
```diff
@@ -0,0 +1,289 @@
+#include "periph/i2c.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG        (1)
```
Disable

aabadie commented:
I addressed your comments @PeterKietzmann

aabadie commented:
@gebart, did you find some time to give this one a try ?

aabadie commented on sys/auto_init/saul/auto_init_tsl2561.c:
```diff
@@ -0,0 +1,75 @@
+                         tsl2561_params[i].addr,
+                         tsl2561_params[i].gain,
+                         tsl2561_params[i].integration) != TSL2561_OK) {
+            LOG_ERROR("[auto_init_saul] error initializing tsl2561 #%u\n", i);
```
For the record, I just (forced) push an update of the logged message (see #6408)

gebart commented:
I can't find my TSL2561 board, resigning from this PR.

aabadie commented:
@gebart, I have [some sensor values displayed on the web](http://riot-demo.inria.fr). The firmware are running RIOT and I added one with the actual code from this PR.
Following the illuminance ranges given by [wikipedia](https://en.wikipedia.org/wiki/Lux), the values raised looks good.

aabadie commented:
@PeterKietzmann, any chance to get this merged ?

PeterKietzmann commented:
I just ordered one (located in Germany). As you may have seen we postponed feature freeze. So probably someone else with that device can test your driver **or** in best case I can, if the device arrives in time.

aabadie commented:
rebased. @PeterKietzmann any chance for you try this one before feature freeze ?

PeterKietzmann commented:
Didn't forget about this one. The device didn't arrive yet but probably it will, today.

aabadie commented:
@PeterKietzmann did you receive the sensor and find time to try this one ?

PeterKietzmann commented:
Received yes, time no. Will solder pinout later. Thanks for the reminder

aabadie commented:
Thanks @PeterKietzmann :)
